bundles/org.openhab.binding.linktap/src/main/java/org/openhab/binding/linktap/internal/PollingDeviceHandler.java

```diff
@@ -0,0 +1,324 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.linktap.internal;
+
+import static org.openhab.binding.linktap.internal.LinkTapBindingConstants.DEVICE_CHANNEL_OH_DURATION_LIMIT;
+import static org.openhab.binding.linktap.internal.LinkTapBindingConstants.DEVICE_CHANNEL_OH_VOLUME_LIMIT;
+import static org.openhab.binding.linktap.protocol.frames.TLGatewayFrame.EMPTY_STRING;
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.validation.constraints.NotNull;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.linktap.configuration.LinkTapDeviceConfiguration;
+import org.openhab.binding.linktap.protocol.frames.DeviceCmdReq;
+import org.openhab.binding.linktap.protocol.frames.TLGatewayFrame;
+import org.openhab.binding.linktap.protocol.http.DeviceIdException;
+import org.openhab.binding.linktap.protocol.http.InvalidParameterException;
+import org.openhab.core.cache.ExpiringCache;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.thing.binding.BridgeHandler;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PollingDeviceHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author David Goodyear - Initial contribution
+ */
+@NonNullByDefault
+public abstract class PollingDeviceHandler extends BaseThingHandler implements IBridgeData {
+
+    private final Logger logger = LoggerFactory.getLogger(PollingDeviceHandler.class);
+
+    protected static final String MARKER_INVALID_DEVICE_KEY = "---INVALID---";
+    protected String registeredDeviceId = EMPTY_STRING;
+
+    private final Object pollLock = new Object();
+
+    protected ExpiringCache<String> lastPollResultCache = new ExpiringCache<>(Duration.ofSeconds(5),
+            PollingDeviceHandler::expireCacheContents);
+
+    private final Object schedulerLock = new Object();
+    private @Nullable ScheduledFuture<?> backgroundGwPollingScheduler;
+
+    private final Object readBackPollLock = new Object();
+    private @Nullable ScheduledFuture<?> readBackPollSf = null;
+
+    protected volatile LinkTapDeviceConfiguration config = new LinkTapDeviceConfiguration();
+
+    protected void requestReadbackPoll() {
+        synchronized (readBackPollLock) {
+            cancelReadbackPoll();
+            scheduler.schedule(() -> {
+                pollForUpdate(true);
+            }, 750, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    protected void cancelReadbackPoll() {
+        synchronized (readBackPollLock) {
```

Comment:
Here and on other places i see you using the `synchronized` keyword and i wonder if all these are essential.
Can you ellaborate on when this lock is occuring/needed ?

Reply:
The places where its synchronizing is literally if different thread's could end up calling the code, to ensure a consistent state, on whats being synchronized. It's mainly the scheduler thread, startup threads / thread calling dispose etc... that from memory overlap. Essentially it's to keep the state consistent in terms of behavior. All locks are scoped by a object responsible for maintaining just the bits its for, to ensure no unexpected side effects.

Reply:
Good you already tried to limit the scope as much as possible. Might be perfectly sane, but i hardly see any binding use these lock that excessive, so i wonder if this is needed at all.
I'll look into this in more detail and might come back to this.

Reply:
I'd prefer to keep these locks in place, this has been soak tested for many months on different VM's and systems during development. I have not seen any issues with them in place. Can these please be left as is?

Reply:
ok